lib/monitor.js

```diff
@@ -42,52 +38,51 @@ class Monitor extends Events.EventEmitter {
             responsePayload: this.settings.responsePayload
         };
 
+        this._reporters = {};
+
         this._ops = this.settings.ops && new Oppsy(server, this.settings.ops.config);
 
         // Event handlers
 
         this._requestLogHandler = (request, event) => {
 
-            this._dataStream.push(new Utils.GreatRequest(request, event));
+            this.push(() => new Utils.GreatRequest(request, event));
         };
 
         this._logHandler = (event) => {
 
-            this._dataStream.push(new Utils.GreatLog(event));
+            this.push(() => new Utils.GreatLog(event));
         };
 
         this._errorHandler = (request, error) => {
 
-            this._dataStream.push(new Utils.GreatError(request, error));
+            this.push(() => new Utils.GreatError(request, error));
         };
 
         this._responseHandler = (request) => {
 
-            this._dataStream.push(new Utils.GreatResponse(request, this._responseOptions, this.settings.filter));
+            this.push(() => new Utils.GreatResponse(request, this._responseOptions, this.settings.filter));
         };
 
         this._opsHandler = (results) => {
 
-            this._dataStream.push(new Utils.GreatOps(results));
+            this.push(() => new Utils.GreatOps(results));
         };
 
         this._wreckHandler = (error, request, response, start, uri) => {
 
-            this._dataStream.push(new Utils.GreatWreck(error, request, response, start, uri));
+            this.push(() => new Utils.GreatWreck(error, request, response, start, uri));
         };
 
-        this._extensionHandler = function (eventName) {
+        this._extensionHandler = function () {
 
-            let event;
-            // (eventName, request, event, tags)
-            if (arguments.length === 4) {
-                event = Object.assign({}, arguments[2]);
-            }
-            else {
-                event = Object.assign({}, arguments[1]);
-            }
-            event.event = eventName;
-            this._dataStream.push(Object.freeze(event));
+            const args = (arguments.length === 1 ? [arguments[0]] : Array.apply(null, arguments));
```

Comment:
Because the way hapi delivers payloads to regular event handlers differs from version to version, this is the only way I could think of to capture all the arguments that hapi sends and still work with version >= 10.x.x. I am open to suggestions here.